keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,532 @@
+managing Endpoints objects for all Services, and a cluster cannot pass
+the conformance test suite unless the Endpoints and EndpointSlice
+Mirroring controllers are running, even though in many cases nothing
+will ever look at the output of the Endpoints controller (and the
+EndpointSlice Mirroring controller will never output anything).
+
+While Kubernetes's API guarantees make it essentially impossible to
+ever actually fully remove Endpoints, we should at least move toward a
+world where most users run Kubernetes with the Endpoints and
+EndpointSlice Mirroring controllers disabled.
+
+## Motivation
+
+### Goals
+
+- Ensure that all documentation reflects the fact that the Endpoints
+  API is deprecated and should not be used by new code, and that
+  Endpoints objects might not be created or mirrored in some clusters.
+
+- Add apiserver warnings when users (other than the Endpoints
+  controller) create Endpoints objects.
+
+- Update the e2e test suite to ensure that all tests that care about
+  endpoints only as a generic concept are using EndpointSlices rather
+  than Endpoints. (For example, the utility function
+  `framework.WaitForServiceEndpointsNum` should check that
+  EndpointSlices exist, not that Endpoints exist).
+
+- Adjust the conformance test suite to not require the Endpoints and
+  EndpointSlice Mirroring controllers to be running:
+
+    - Fix tests such as `Service endpoints latency should not be very
+      high` to check the latency of EndpointSlices rather than
+      Endpoints, since that is what is actually important in
+      real-world usage.
+
+    - Demote tests such as `EndpointSliceMirroring should mirror a
+      custom Endpoints resource through create update and delete` to
+      non-conformance.
+
+    - Split up tests such as `EndpointSlice should create Endpoints
+      and EndpointSlices for Pods matching a Service` into a
+      conformance test for EndpointSlice and a non-conformance test
+      for Endpoints.
+
+- Update the e2e test suite so that all remaining tests that depend on
+  the outputs of the Endpoints or EndpointSlice Mirroring controllers
+  are feature-tagged so they can be skipped in clusters that do not
+  run those controllers, and set up a periodic e2e job running in such
+  a configuration.
+
+- Explicitly document that disabling `endpoints-controller` and/or
+  `endpointslice-mirroring-controller` via kube-controller-manager's
+  `--controllers` flag is a supported and conforming configuration.
+
+    - Figure out what to do with stale Endpoints in such a cluster.
+
+```
+<<[UNRESOLVED kubernetes.default ]>>
+
+- MAYBE change kube-apiserver to optionally not generate Endpoints for
+  kubernetes.default, though this would require adding a
+  kube-apiserver configuration option, and the benefit from removing
+  just that 1 Endpoints object is small. Perhaps instead it could just
+  add an annotation to the object pointing out the fact that Endpoints
+  is deprecated.
```

Comment:
From experience, I can say that it would make teaching easier if there were an EndpointSlice for the `kubernetes` Service in the default namespace.

If we deprecate, we should mean it.

Reply:
Oh, it already creates EndpointSlices. (It has to, because kube-proxy doesn't look at Endpoints.)

The point here is that, if we consider the existence of a `default/kubernetes` Endpoints object to be an "API", then kube-apiserver's default behavior has to be to continue generating both Endpoints and EndpointSlices for it, and in that case, there's not a huge benefit to adding a command-line option to tell it to stop generating the Endpoints.

(If people think that the `default/kubernetes` Endpoints is _not_ an API, then we could just eventually stop generating it, without needing a config option.)